packages/eslint-plugin/src/rules/no-unused-vars.ts

```diff
@@ -642,6 +672,41 @@ export default createRule<Options, MessageIds>({
       'Program:exit'(programNode): void {
         const unusedVars = collectUnusedVariables();
 
+        /**
+         * Collect unused import specifiers for the import declaration.
+         * Access the map by the import declaration.
+         */
+        const unusedImportSpecifiersMap = new WeakMap<
```

Comment:
[Style] A few interlinked suggestions:

* This comment doesn't add anything IMO: the variable is already named "unused import specifiers", and its type says that import declarations are the keys
* The `Map` suffix in the name isn't necessary either - it's a map type already
* It's kind of confusing to read through so much new logic, where a few dozen lines are just to populate a map. Let's move the logic into a function outside of the rule.

```suggestion
        const unusedImportSpecifiers =
          collectUnusedImportSpecifiers(unusedVars);
```